### ISSUE TITLE
Function declaration in inner function becomes undefined within a loop

### ISSUE THREAD

**Issue by kpreisser** [paulbartrum/jurassic#140]
Hi, I encountered a strange bug today:

```c#
        var engine = new ScriptEngine();
        engine.SetGlobalValue("console", new FirebugConsole(engine));

        engine.Execute(@"

'use strict';
for (var i = 0; i < 3; i++) {
    (function () {
        function test() {
        }
        console.log(typeof test);
    }());
}

");
```

**Expected output:**
```
function
function
function
```

**Actual output:**
```
function
undefined
undefined
```

I was surprised that this was not caught/tested by the unit tests. I have not yet looked what could be the cause that the `test` function is undefined after the first iteration.

If you change the decaration of `test` to be a function expression that is assigned to a variable, it works:
```javascript
for (var i = 0; i < 3; i++) {
    (function () {
        var test = function test() {
        };
        console.log(typeof test);
    }());
}
```

Also, if you extract the first inner function outside of the loop, the output is correct:
```javascript
function loopTest() {
    var test = function test() {
    };
    console.log(typeof test);
}

for (var i = 0; i < 3; i++)
    loopTest();
```

Thanks!

**Comment by kpreisser** [comment]
It seems the problem arises due to a combination of two factors:
1. The `LoopStatement` generates the code for its body (as well as the initializer and the condition) **two times**, as described in the `GenerateCode()` method:
https://github.com/paulbartrum/jurassic/blob/e8ade10c448ccd99ca40217bf6221c89259f1d9f/Jurassic/Compiler/Statements/LoopStatement.cs#L115-L130

2. `Scope.GenerateDeclarations()` checks if a variable has already been initialized, and in that case skips it:
https://github.com/paulbartrum/jurassic/blob/e8ade10c448ccd99ca40217bf6221c89259f1d9f/Jurassic/Compiler/Scope/Scope.cs#L263-L265

However, when the `LoopStatement` generates the code for the body the second time, `Scope.GenerateDeclarations()` determines that the variable (`test`)  has already been initialized, and therefore skips code generation to create a new `FunctionInstance` and store it into the `test` variable. This results in `test` being undefined except for the first loop iteration.

What would be the right fix for this issue? Change the `LoopStatement` so that it generates the code for the `<initializer>`, `<condition>` and `<loop body>` only once?

**Comment by paulbartrum** [comment]
Thanks for this. I ended up modifying your fix quite a bit, check out commit 875a93f for details.

**Comment by kpreisser** [comment]
Thank you!
